tests/test_hypersphere.py

```diff
@@ -140,23 +174,100 @@ def test_exp_vectorization(self):
         self.assertTrue(gs.allclose(result.shape, (n_samples, dim + 1)))
+        expected = gs.zeros((n_samples, dim + 1))
+        for i in range(n_samples):
+            expected[i] = self.metric.exp(n_tangent_vecs[i], n_base_points[i])
+        self.assertTrue(gs.allclose(result, expected))
+
+    def test_exp_vectorization_with_depth(self):
+        n_samples = self.n_samples
+        depth = self.depth
+
+        dim = self.dimension
+        one_vec = self.space.random_uniform(
+            n_samples=1, depth=depth)
+        one_base_point = self.space.random_uniform(
+            n_samples=1, depth=depth)
+        n_vecs = self.space.random_uniform(
+            n_samples=n_samples, depth=depth)
+        n_base_points = self.space.random_uniform(
+            n_samples=n_samples, depth=depth)
+
+        one_tangent_vec = self.space.projection_to_tangent_space(
+            one_vec, base_point=one_base_point)
+        result = self.metric.exp(one_tangent_vec, one_base_point)
+        self.assertTrue(
+            gs.allclose(result.shape, (depth, dim + 1)),
+            'result.shape = {}'.format(result.shape))
+        expected = gs.zeros((depth, dim + 1))
+        for j in range(depth):
+            expected[j] = self.metric.exp(
+                one_tangent_vec[j], one_base_point[j])
+        self.assertTrue(gs.allclose(result, expected))
+
+        n_tangent_vecs = self.space.projection_to_tangent_space(
+            n_vecs, base_point=one_base_point)
+        result = self.metric.exp(n_tangent_vecs, one_base_point)
+        self.assertTrue(
+            gs.allclose(result.shape, (n_samples, depth, dim + 1)),
+            '\n result.shape = {}'.format(result.shape))
+        expected = gs.zeros((n_samples, depth, dim + 1))
+        for i in range(n_samples):
+            for j in range(depth):
+                expected[i, j] = self.metric.exp(
+                    n_tangent_vecs[i, j], one_base_point[j])
+        self.assertTrue(gs.allclose(result, expected))
+
+        one_tangent_vec = self.space.projection_to_tangent_space(
+            one_vec, base_point=n_base_points)
+        result = self.metric.exp(one_tangent_vec, n_base_points)
+        self.assertTrue(
+            gs.allclose(result.shape, (n_samples, depth, dim + 1)))
+        expected = gs.zeros((n_samples, depth, dim + 1))
+        for i in range(n_samples):
+            for j in range(depth):
+                expected[i, j] = self.metric.exp(
+                    one_tangent_vec[i, j], n_base_points[i, j])
+        self.assertTrue(gs.allclose(result, expected))
+
+        n_tangent_vecs = self.space.projection_to_tangent_space(
+            n_vecs, base_point=n_base_points)
+        result = self.metric.exp(n_tangent_vecs, n_base_points)
+        self.assertTrue(
+            gs.allclose(result.shape, (n_samples, depth, dim + 1)))
+        expected = gs.zeros((n_samples, depth, dim + 1))
+        for i in range(n_samples):
```

Comment:
can you do this w/o the two nested for loops?

Reply:
Yes, we could vectorize it. But we want to unit test the vectorization here, so I think we should keep it?